VirtualHavruta/vh.py

```diff
@@ -346,7 +356,190 @@ def retrieve_docs(self, query: str, msg_id: str='', filter_mode: str='primary'):
+            query = f"""
+            MATCH (start {{`metadata.url`: $url}})
+            WITH start
+            MATCH (start){start_node_operator}[:FROM_TO*{i}]{related_node_operator}(neighbor)
+            WHERE neighbor <> start
+            {source_filter}
+            RETURN DISTINCT neighbor, {i} AS depth
+            """
+            with GraphDatabase.driver(self.config["database"]["kg"]["url"], auth=(self.config["database"]["kg"]["username"], self.config["database"]["kg"]["password"])) as driver:
+                neighbor_nodes, _, _ = driver.execute_query(
+                query,
+                parameters_=query_params,
+                database_=self.config["database"]["kg"]["name"],)
+            nodes.extend(neighbor_nodes)
+        return nodes
+    
+    def query_node_by_url(self, url: str,) -> str|None:
+        """Given a url, query the graph database for the node with that url.
+
+        If more than one node has the same url, return only one.
+
+        Parameters
+        ----------
+        url
+            of node
+
+        Returns
+        -------
+            unique id of the node
+        """
+        query_parameters = {"url": url}
+        query_string="""
+        MATCH (n)
+        WHERE n.`metadata.url`=$url
+        RETURN n.id
+        LIMIT 1
+        """
+        with GraphDatabase.driver(self.config["database"]["kg"]["url"], auth=(self.config["database"]["kg"]["username"], self.config["database"]["kg"]["password"])) as driver:
+            id, _, _ = driver.execute_query(
+            query_string,
+            parameters_=query_parameters,
+            database_=self.config["database"]["kg"] ["name"],)
+        if id:
+            return id[0].data()["n.id"]
+        else:
+            return None
+
+    def query_graph_db_by_url(self, urls: list[str]) -> list[Document]:
+        """Given a list of urls, query the graph database for the nodes with those urls.
+
+        Note that there is a one-to-many relationship between urls and documents in the vector database,
+        due to different sources for the same url.
+
+        Return the nodes as document compatible type.
+
+        Parameters
+        ----------
+        urls
+            of documents
+
+        Returns
+        -------
+            list of documents
+        """
+        query_parameters = {"urls": urls}
+        query_string="""
+        MATCH (n)
+        WHERE n.`metadata.url` in $urls
+        RETURN n
+        """
+        with GraphDatabase.driver(self.config["database"]["kg"]["url"], auth=(self.config["database"]["kg"]["username"], self.config["database"]["kg"]["password"])) as driver:
+            nodes, _, _ = driver.execute_query(
+            query_string,
+            parameters_=query_parameters,
+            database_=self.config["database"]["kg"]["name"],)
+        return [convert_node_to_doc(node) for node in nodes]
+
+    def sort_reference(self, query: str, retrieval_res, msg_id: str = ''):
```

Comment:
We still need primary/secondary filters. Please incorporate that still.

Reply:
Readded the parameter filter_mode.